core/server/master/src/test/java/alluxio/master/meta/InodeLockManagerTest.java

```diff
@@ -0,0 +1,89 @@
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.meta;
+
+import static org.junit.Assert.assertFalse;
+
+import alluxio.master.file.meta.Edge;
+import alluxio.master.file.meta.InodeFile;
+import alluxio.master.file.meta.InodeLockManager;
+import alluxio.master.file.meta.InodeTree.LockMode;
+import alluxio.master.file.options.CreateFileOptions;
+import alluxio.resource.LockResource;
+import alluxio.util.CommonUtils;
+
+import org.junit.Test;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Unit tests for {@link InodeLockManager}.
+ */
+public class InodeLockManagerTest {
+  @Test(timeout = 10000)
+  public void lockInode() throws Exception {
+    inodeLockTest(LockMode.WRITE, LockMode.READ, true);
+    inodeLockTest(LockMode.READ, LockMode.WRITE, true);
+    inodeLockTest(LockMode.WRITE, LockMode.WRITE, true);
+    inodeLockTest(LockMode.READ, LockMode.READ, false);
+  }
+
+  @Test(timeout = 10000)
+  public void lockEdge() throws Exception {
+    edgeLockTest(LockMode.WRITE, LockMode.READ, true);
+    edgeLockTest(LockMode.READ, LockMode.WRITE, true);
+    edgeLockTest(LockMode.WRITE, LockMode.WRITE, true);
+    edgeLockTest(LockMode.READ, LockMode.READ, false);
+  }
+
+  private void inodeLockTest(LockMode take, LockMode tryToTake, boolean expectBlocking)
+      throws Exception {
+    InodeLockManager lockManager = new InodeLockManager();
+    AtomicBoolean threadFinished = new AtomicBoolean(false);
+    InodeFile inode = InodeFile.create(0, 0, "name", 0, CreateFileOptions.defaults());
+    LockResource lock = lockManager.lockInode(inode, take);
+    Thread t = new Thread(() -> {
+      // Copy the inode to make sure we aren't comparing inodes by reference.
+      InodeFile inodeCopy = InodeFile.fromJournalEntry(inode.toJournalEntry().getInodeFile());
+      try (LockResource lr = lockManager.lockInode(inodeCopy, tryToTake)) {
+        threadFinished.set(true);
+      }
+    });
+    t.start();
+    if (expectBlocking) {
+      CommonUtils.sleepMs(20);
+      assertFalse(threadFinished.get());
+      lock.close();
+    }
+    CommonUtils.waitFor("lock to be acquired by the second thread", () -> threadFinished.get());
+  }
+
+  private void edgeLockTest(LockMode take, LockMode tryToTake, boolean expectBlocking)
+      throws Exception {
+    InodeLockManager lockManager = new InodeLockManager();
+    AtomicBoolean threadFinished = new AtomicBoolean(false);
+    LockResource lock = lockManager.lockEdge(new Edge(10, "name"), take);
+    Thread t = new Thread(() -> {
+      // Use a new Edge each time to make sure we aren't comparing edges by reference.
+      try (LockResource lr = lockManager.lockEdge(new Edge(10, "name"), tryToTake)) {
+        threadFinished.set(true);
+      }
+    });
+    t.start();
+    if (expectBlocking) {
+      CommonUtils.sleepMs(20);
```

Comment:
are these waits long enough even when there are many threads/tests running concurrently, on slower machines?

Reply:
Running locally it takes less than 1ms for the second thread to try taking the lock, so the test fails consistently if I break the locking code. I don't want to make the sleeps longer because it will slow down the test. Incorrect code may sometimes pass the test on slow/overloaded machines, but we will never have a flaky failure.